subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c

```diff
@@ -538,3 +569,306 @@ int nrf_cloud_decode_data_endpoint(const struct nrf_cloud_data *input,
+		   (mod_inf->device == NRF_CLOUD_INFO_SET)) {
+		LOG_ERR("CONFIG_MODEM_INFO_ADD_DEVICE is not enabled, unable to add device info");
+		return -EACCES;
+	} else if ((!IS_ENABLED(CONFIG_MODEM_INFO_ADD_NETWORK)) &&
+		   (mod_inf->network == NRF_CLOUD_INFO_SET)) {
+		LOG_ERR("CONFIG_MODEM_INFO_ADD_NETWORK is not enabled, unable to add network info");
+		return -EACCES;
+	} else if ((!IS_ENABLED(CONFIG_MODEM_INFO_ADD_SIM)) &&
+		   (mod_inf->sim == NRF_CLOUD_INFO_SET)) {
+		LOG_ERR("CONFIG_MODEM_INFO_ADD_SIM is not enabled, unable to add SIM info");
+		return -EACCES;
+	}
+
+	int err = 0;
+	cJSON *tmp = cJSON_CreateObject();
+
+	if (!tmp) {
+		err = -ENOMEM;
+		goto cleanup;
+	}
+
+#ifdef CONFIG_MODEM_INFO
+	struct modem_param_info *mpi = (struct modem_param_info *)mod_inf->mpi;
+	struct modem_param_info fetched_mod_inf;
+
+	if (!mpi) {
+		err = modem_info_init();
+		if (err) {
+			LOG_ERR("modem_info_init() failed: %d", err);
+			goto cleanup;
+		}
+
+		err = modem_info_params_init(&fetched_mod_inf);
+		if (err) {
+			LOG_ERR("modem_info_params_init() failed: %d", err);
+			goto cleanup;
+		}
+
+		err = modem_info_params_get(&fetched_mod_inf);
+		if (err < 0) {
+			LOG_ERR("modem_info_params_get() failed: %d", err);
+			goto cleanup;
+		}
+		mpi = &fetched_mod_inf;
+	}
+
+	err = modem_info_json_object_encode(mpi, tmp);
+	if (err < 0) {
+		LOG_ERR("Failed to encode modem info: %d", err);
+		goto cleanup;
+	}
+	err = 0;
+#endif
+
+	if (encode_info_item(mod_inf->device, MODEM_INFO_JSON_KEY_DEV_INF, tmp, mod_inf_obj) ||
+	    encode_info_item(mod_inf->network, MODEM_INFO_JSON_KEY_NET_INF, tmp, mod_inf_obj) ||
+	    encode_info_item(mod_inf->sim, MODEM_INFO_JSON_KEY_SIM_INF, tmp, mod_inf_obj)) {
+		LOG_ERR("Failed to encode modem info");
+		err = -EIO;
+		goto cleanup;
+	}
+
+cleanup:
+	cJSON_Delete(tmp);
+	return err;
+}
+
+int nrf_cloud_service_info_json_encode(const struct nrf_cloud_svc_info *const svc_inf,
+	cJSON *const svc_inf_obj)
+{
+	if (!svc_inf || !svc_inf_obj) {
+		return -EINVAL;
+	}
+
+	int err = nrf_cloud_encode_service_info_fota(svc_inf->fota, svc_inf_obj);
+
+	if (err) {
+		return err;
+	}
+
+	return nrf_cloud_encode_service_info_ui(svc_inf->ui, svc_inf_obj);
+}
+
+void nrf_cloud_device_status_free(struct nrf_cloud_data *status)
+{
+	if (status && status->ptr) {
+		cJSON_free((void *)status->ptr);
+		status->ptr = NULL;
+		status->len = 0;
+	}
+}
+
+int nrf_cloud_device_status_encode(const struct nrf_cloud_device_status *const dev_status,
+	struct nrf_cloud_data * const output)
+{
+	if (!dev_status || !output) {
+		return -EINVAL;
+	}
+
+	int err = 0;
+	cJSON *root_obj = cJSON_CreateObject();
+	cJSON *state_obj = cJSON_AddObjectToObjectCS(root_obj, JSON_KEY_STATE);
+	cJSON *reported_obj = cJSON_AddObjectToObjectCS(state_obj, JSON_KEY_REP);
+	cJSON *device_obj = cJSON_AddObjectToObjectCS(reported_obj, JSON_KEY_DEVICE);
+	cJSON *svc_inf_obj = cJSON_AddObjectToObjectCS(device_obj, JSON_KEY_SRVC_INFO);
+
+	if (svc_inf_obj == NULL) {
```

Comment:
That works on the assumption that the thread who executes this function is not preempted. I can imagine a scenario that we fail to allocate some of the previous objects, get preempted and the other thread frees some memory. In such case the final allocation may succeed and we end up with malformed object. Not very likely to happen, but still, possible.

Up to you to decide whether sendin an incomplete json is acceptable for this case.

Reply:
i had not considered such a scenario, but i don't think that it applies to this specific case due to how the JSON is nested.  Each item is allocated and added into the previous item.  If there is a failure, the next one will also fail because it can't add the item to a NULL obj.

i will review the rest of this file to see if there are any cases where the preemption scenario would cause problems.

Reply:
i didn't find any preemption issues, but i did add a commit to clean up a lot of the cJSON handling.